### DISCUSSION TITLE
Issue with Undo/Redo in superBrewer3000.notation

### DISCUSSION THREAD

**Discussion by amiths89** [eclipse-emfcloud/emfcloud#235] [Coffee Editor]
I am currently working on emfcloud/modelserver-glsp-integration which I understand is under maintenance mode so I did run into troubles to get it to work but I eventually got it running. But now I see an issue in it which seems to be an issue with Coffee Editor as well. The undo/redo seems to not work here in Coffee Editor as well when I try to move a node in superBrewer3000.notation and then do an undo, it does not move back unless I refresh the page or do any event such as create a new node or delete an existing node.

I have noticed this issue in modelserver-glsp-integration where the EMSSubscriptionListener refreshes for increamentalUpdates but not for onDirtyChange. And so here we can see that undo/redo is not an increamentalUpdate and instead it is a dirtyState which is why it fails to refresh for undo/redo. Will there be a fix for this issue or is there a way to get around this?

**Comment by eneufeld** [comment]
Hi @amiths89 ,
I need to check this. Thank you for the detailed steps. I wonder why we have a dirty notification but no change notification.
Anyway you should be able to replace the Listener:
Use your custom EMSSourceModelStorage by binding it in EMSGLSPDiagramModule.java#bindSourceModelStorage
Override createSubscriptionListener in EMSSourceModelStorage

This should be enough.

**Reply by amiths89** [reply]
@eneufeld Thanks for the solution. I will try this.
